### PR TITLE
GTEST/UCT: Unit tests for UMR XGVMI

### PR DESCRIPTION
**Pull request by iyastreb**
## What
Few unit tests for XGVMI UMR approach to check that:
- Lazy initialization works as expected
- All UMR invariants are respected
- UMR mkey pool is properly used on exporter when XGVMI is supported and when it's not
- UMR mkey hash map is properly used on importer side with and without XGVMI support
- All allocated objects are freed

We introduce new config option UCX_XGVMI_UMR_ENABLE option as a part of this commit, which enables XGVMI UMR workflow.

### CODE REVIEW

File: test/gtest/uct/ib/test_devx.cc [hunk 1]
```diff
@@ -48,4 +48,202 @@ UCS_TEST_P(test_devx, dbrec)
...
 }
 
+class test_devx_umr_mkey : public test_devx {
+public:
+    void init() {
+        test_devx::init();
+    }
+
+    uct_ib_mlx5_devx_mem_t *create_memh(size_t length) const {
+        void *addr                     = ucs_malloc(length, "test");
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:58]: sure

File: test/gtest/uct/ib/test_devx.cc [hunk 2]
```diff
@@ -48,4 +48,202 @@ UCS_TEST_P(test_devx, dbrec)
...
+    void destroy_memh(uct_ib_mlx5_devx_mem_t *memh) const {
+        void *addr                       = memh->address;
+        uct_md_mem_dereg_params_t params = {};
+
+        params.field_mask = UCT_MD_MEM_DEREG_FIELD_MEMH;
+        params.memh       = (uct_mem_h)memh;
+
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_dereg(m_e->md(), &params));
+
+        if (nullptr != addr) {
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:76]: no, for imported memory handles it's NULL

File: test/gtest/uct/ib/test_devx.cc [hunk 3]
```diff
@@ -48,4 +48,202 @@ UCS_TEST_P(test_devx, dbrec)
...
+    check_memh_export(memh, true);
+    destroy_memh(memh);
+
+    /* After release of memh, associated UMR mkey is moved back to mkey pool */
+    ASSERT_EQ(1, ucs_list_length(&md()->umr.mkey_pool));
+}
+
+UCS_TEST_P(test_devx_umr_mkey, export_mkey_xgvmi_reuse)
+{
+    if (!(md()->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI)) {
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:144]: Yes, indeed that is a good option!
I tried that initially, but if failed with exception, because `m_e` was created only in `init`, but this check-for-skip-test is called after object constructor but before `init`. So I moved `m_e` object creation to the ctor and it works like that

File: test/gtest/uct/ib/test_devx.cc [hunk 4]
```diff
@@ -48,4 +48,202 @@ UCS_TEST_P(test_devx, dbrec)
...
+
+UCS_TEST_P(test_devx_umr_mkey, import_mkey_no_umr)
+{
+    /* Export any mkey */
+    uct_ib_mlx5_devx_mem_t *exported_memh = create_memh(1024);
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), exported_memh));
+
+    /* Construct packed mkey without UMR tag */
+    uct_ib_md_packed_mkey_t packed_mkey = {
+        .lkey = exported_memh->exported_lkey & ~UCT_IB_MLX5_MKEY_TAG_UMR,
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:192]: I tested it manually, but I was advised by @brminich to not commit such a large buffer allocation (2GB), because it may slow down / break the tests.

File: test/gtest/uct/ib/test_devx.cc [hunk 5]
```diff
@@ -48,4 +48,202 @@ UCS_TEST_P(test_devx, dbrec)
...
+    ASSERT_EQ(MEMH_COUNT - REUSE_COUNT, ucs_list_length(&md()->umr.mkey_pool));
+
+    for (size_t i = 0; i < REUSE_COUNT; ++i) {
+        destroy_memh(memh[i]);
+    }
+
+    ASSERT_EQ(MEMH_COUNT, ucs_list_length(&md()->umr.mkey_pool));
+}
+
+UCS_TEST_P(test_devx_umr_mkey, import_mkey_no_umr)
```
Comment (tvegas1) [inline test/gtest/uct/ib/test_devx.cc:184]: seems that test is failing on `ALLOW_OTHER_VHCA_ACCESS` on non-supported xgvmi FW, called from `uct_ib_mlx5_devx_xgvmi_umem_mr()`.

We might need to fix it in separate PR with either enabling `silent` or removing `uct_ib_mlx5_devx_allow_xgvmi_access()` when non supported.

File: test/gtest/uct/ib/test_devx.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:144]: Yes, this is exactly what I've implemented

[the remark this answers is not in the document]
Comment (tvegas1) [reply test/gtest/uct/ib/test_devx.cc:144]: i see, i misunderstood you last comment.

File: test/gtest/uct/ib/test_devx.cc [hunk 6]
```diff
@@ -48,4 +48,202 @@ UCS_TEST_P(test_devx, dbrec)
...
+    ASSERT_NE(imported_memh, imported_memh2);
+    ASSERT_EQ(imported_memh->super.lkey, imported_memh2->super.lkey);
+    ASSERT_EQ(1, kh_size(md()->umr.mkey_hash));
+
+    destroy_memh(imported_memh);
+    destroy_memh(imported_memh2);
+    destroy_memh(exported_memh);
+}
+
 UCT_INSTANTIATE_IB_TEST_CASE(test_devx);
```
Comment (brminich) [inline test/gtest/uct/ib/test_devx.cc:263]: pls move it before the new class, typically we initialize test classes right after their (and their tests) definitions

Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:263]: sure

File: test/gtest/uct/ib/test_devx.cc [hunk 7]
```diff
@@ -48,17 +50,42 @@ UCS_TEST_P(test_devx, dbrec)
...
+    }
+
+    /*
+     * XGVMI capability advertised by device might be dropped on the initial
+     * operation. This function performs dummy XGVMI operation to detect
+     * whether XGVMI is actually operational
+     */
+    bool check_xgvmi_effective() const {
+        if (check_xgvmi()) {
+            uct_ib_mlx5_devx_mem_t *memh = create_memh(1);
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:73]: This change would have some side-effects:
- `check_xgvmi_effective` calls `reg_export` which internally initializes UMR export objects, and populates mkey pool with one element. So after this call we would not be able to check the pristine state (like we do in `lazy_init` test)
- with this change we would not be able to test cases where XGVMI capability is advertised, but not actually supported. Currently I have one test that covers this use case: `export_mkey_xgvmi`. I think it's desirable to have a possibility to test it

So I would keep it as is, calling this function only where necessary

**Review by brminich — DISMISSED** [review]

**Review by tvegas1 — APPROVED** [review]

File: test/gtest/uct/ib/test_devx.cc [hunk 8]
```diff
@@ -16,12 +16,14 @@ class test_devx : public uct_test {
 public:
     entity* m_e;
 
+    test_devx() : m_e(create_entity(0))
+    {
+        m_entities.push_back(m_e);
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:21]: IMO it better be during init() and not in actor

File: test/gtest/uct/ib/test_devx.cc [hunk 9]
```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
...
+    bool check_xgvmi() const {
+        return md()->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI;
+    }
+
+    /*
+     * XGVMI capability advertised by device might be dropped on the initial
+     * operation. This function performs dummy XGVMI operation to detect
+     * whether XGVMI is actually operational
+     */
+    bool check_xgvmi_effective() const {
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:71]: can we unite check_xgvmi_effective and check_xgvmi to a single function?

File: test/gtest/uct/ib/test_devx.cc [hunk 10]
```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
...
+    uct_ib_mlx5_devx_mem_t *memh = create_memh(1024);
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh));
+    /* XGVMI capability might be dropped by this point! */
+
+    /* UMR export transport objects must be initialized */
+    check_umr_init(true, false);
+    /* UMR mkey pool must be empty */
+    ASSERT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
+    /* Whether mkey is UMR depends on XGVMI capability */
+    check_memh_export(memh, check_xgvmi());
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:157]: function is skipped if check_xgvmi is false. so why. check it here again? (multiple locations)

File: test/gtest/uct/ib/test_devx.cc [hunk 11]
```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
...
+    void destroy_memh(uct_ib_mlx5_devx_mem_t *memh) const {
+        void *addr                       = memh->address;
+        uct_md_mem_dereg_params_t params = {};
+
+        params.field_mask = UCT_MD_MEM_DEREG_FIELD_MEMH;
+        params.memh       = (uct_mem_h)memh;
+
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_dereg(m_e->md(), &params));
+
+        if (nullptr != addr) {
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:103]: ucs_free ignores NULL anyway, no need to check

File: test/gtest/uct/ib/test_devx.cc [hunk 12]
```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
...
+
+        EXPECT_NE(nullptr, addr);
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_reg(m_e->md(), addr, length, &params,
+                                               &memh));
+        return ucs_derived_of(memh, uct_ib_mlx5_devx_mem_t);
+    }
+
+    void destroy_memh(uct_ib_mlx5_devx_mem_t *memh) const {
+        void *addr                       = memh->address;
+        uct_md_mem_dereg_params_t params = {};
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:96]: why need to init?

File: test/gtest/uct/ib/test_devx.cc [hunk 13]
```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
...
+
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_dereg(m_e->md(), &params));
+
+        if (nullptr != addr) {
+            ucs_free(addr);
+        }
+    }
+
+    void check_umr_init(bool init_export, bool init_import) const {
+        EXPECT_EQ(init_export, nullptr != md()->umr.qp);
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:109]: if (init_export) {
  EXPECT_NE(nullptr, umr.qp);
  EXPECT_NE(nullptr, umr.cq);
} else {
   EXPECT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
}

File: test/gtest/uct/ib/test_devx.cc [hunk 14]
```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
...
+        EXPECT_EQ(init_export, nullptr != md()->umr.cq);
+        if (!init_export) {
+            EXPECT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
+        }
+
+        EXPECT_EQ(init_import, nullptr != md()->umr.mkey_hash);
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:115]: seems redundant, we initialize the hash on demand anyway, and it's implementation detail

File: test/gtest/uct/ib/test_devx.cc [hunk 15]
```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
...
+    }
+
+    static void check_memh_export(uct_ib_mlx5_devx_mem_t *memh, bool is_umr) {
+        EXPECT_EQ(is_umr, UCT_IB_MLX5_MKEY_TAG_UMR ==
+                          (memh->exported_lkey & 0xff));
+        EXPECT_EQ(is_umr, nullptr != memh->exported_umr_mkey);
+        EXPECT_EQ(is_umr, nullptr == memh->cross_mr);
+    }
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:123]: I think this check is too low-level, we should check import/export works under certain conditions, no need to check implementation details such as tag or struct fields
too low-level checks will make it harder to refactor the code while preserving the UCT level functionality

File: test/gtest/uct/ib/test_devx.cc [hunk 16]
```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
...
+UCS_TEST_SKIP_COND_P(test_devx_umr_mkey, export_mkey_xgvmi, !check_xgvmi())
+{
+    uct_ib_mlx5_devx_mem_t *memh = create_memh(1024);
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh));
+    /* XGVMI capability might be dropped by this point! */
+
+    /* UMR export transport objects must be initialized */
+    check_umr_init(true, false);
+    /* UMR mkey pool must be empty */
+    ASSERT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:155]: why is it important the mkey pool is empty?

File: test/gtest/uct/ib/test_devx.cc [hunk 17]
```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
...
+    /* Whether mkey is UMR depends on XGVMI capability */
+    check_memh_export(memh, check_xgvmi());
+
+    destroy_memh(memh);
+
+    /* After release of memh, associated UMR mkey is moved back to mkey pool */
+    ASSERT_EQ(check_xgvmi() ? 1 : 0, ucs_list_length(&md()->umr.mkey_pool));
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:162]: why is it needed to check?

File: test/gtest/uct/ib/test_devx.cc [hunk 18]
```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
...
+}
+
+UCS_TEST_SKIP_COND_P(test_devx_umr_mkey, export_mkey_xgvmi_reuse,
+                     !check_xgvmi_effective())
+{
+    const size_t MEMH_COUNT = 10;
+    uct_ib_mlx5_devx_mem_t *memh[MEMH_COUNT];
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:169]: maybe std::vector?

File: test/gtest/uct/ib/test_devx.cc [hunk 19]
```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
...
+    }
+
+    /* After release all allocated UMR keys go back to the mkey pool */
+    ASSERT_EQ(MEMH_COUNT, ucs_list_length(&md()->umr.mkey_pool));
+
+    /* Now reuse the keys for different handles of different size */
+    const size_t REUSE_COUNT = 5;
+    for (size_t i = 0; i < REUSE_COUNT; ++i) {
+        memh[i] = create_memh(2048);
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh[i]));
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:191]: can we check performance or somehow check there is no FW flow when reusing an existing key?
checking mkey_pool list length is too low level and indirect

File: test/gtest/uct/ib/test_devx.cc [hunk 20]
```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
...
+                     !check_xgvmi_effective())
+{
+    /* Export any mkey */
+    uct_ib_mlx5_devx_mem_t *exported_memh = create_memh(1024);
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), exported_memh));
+
+    /* Construct packed mkey without UMR tag */
+    uct_ib_md_packed_mkey_t packed_mkey = {
+        .lkey = exported_memh->exported_lkey & ~UCT_IB_MLX5_MKEY_TAG_UMR,
+        .vhca_id = md()->super.vhca_id
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:214]: can we disable UMR export in another way? maybe by configuration?
it's too low-level to use mkey tags in the test

**Review by yosefe — COMMENTED** [review]
as general comment, make the tests more high level to capture more flows in less tests

File: test/gtest/uct/ib/test_devx.cc (hunk 19 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:191]: Good question!
Regarding checking FW flow - that's a classical case for mocking, we could easily do this if we used mocking library/framework (that we discussed some time ago). Then we would write something like:
`mock().expectNcalls(0, "uct_ib_mlx5_devx_obj_create");`
But without mocking framework I've no idea how we can check that.

wrt check performance - is it fast and reliable way to guarantee that FW code is not triggered? I don't think so..
So IMO the only fast and reliable way to check this requirement is to check the pool size

File: test/gtest/uct/ib/test_devx.cc (hunk 20 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:214]: These are the options that I see:
1. Add configuration e.g. `UCX_UMR_ENABLE` like you said. IMO this is the least desirable option, because there is no use case in production workflow for this option (right?), so we add it in production code purely for unit testing purposes, which seems wrong.
2. Using mocking library. We would be able to mock internal function call like:
`mock().expectOneCall("uct_ib_mlx5_is_xgvmi_supported").andReturnValue(0);`
3. Create MR of size > 2GB, in this case it will be handled by KSM workflow. Mikhail advised me not to do that in unit test, because it can be an expensive call.
4. Remove UMR tag from the mkey - like is done today. IMO it's ok to do low-level things in unit tests - they are low-level tests by definition.

Do you have a preference?

File: test/gtest/uct/ib/test_devx.cc (hunk 15 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:123]: There are several comments mentioning low-level, so let me respond here.
Of course it's low level by definition, these are unit tests, not integration ones.
First I agree with you that having excessive amount of unit tests, that check implementation, not behaviour, can be hard to maintain. And ROI of those tests can be negative. I believe it's not the case here.

> we should check import/export works under certain conditions

We already have high-level CI tests for XGVMI that check that, right? There are however many things that we cannot check easily with integration tests. The fact that "it works" does not necessarily mean that it works correctly. I give you a few real examples when integration test "works":
- How do we make sure that UMR approach is used, not KSM? You cannot reliably verify that at high-level
- How to make sure that mkeys are actually reused and stored in the pool? You don't see this at high level
- Same for importer side - check that we don't create new alias each time, but reuse one from hash map
- Check that we don't allocate objects unless we need to
- Finally it takes much longer to run integration tests, so we cannot test all the possible corner cases

The aim is to complement integration tests with the reasonable amount of low-level tests. with the intention to:
1. Code feature requirements and assumptions, so that they exist not just in the design documents or developers' heads.
2. Execute different workflows to check all the internal invariants and assertions
3. Provide the minimum set of tests that cover all flows and requirements, and are executed instantly.
4. Provide a safety net to actually make a refactoring **easier**. It's super hard to change the code that is not covered by tests sufficiently. IMO it's better to break and fix a unit test than code in production

File: test/gtest/uct/ib/test_devx.cc (hunk 9 above)
Comment (yosefe) [reply test/gtest/uct/ib/test_devx.cc:71]: why do we care about testing "fake-gvmi" devices? IMO we should check if xgvmi is "really" supported (or not)

File: test/gtest/uct/ib/test_devx.cc (hunk 10 above)
Comment (yosefe) [reply test/gtest/uct/ib/test_devx.cc:154]: i'd suggest that check_xgvmi will trigger some flow that checks if xgvmi is really supported. or remove it altogether and add a skip during the test itself if xgvmi is dropped

File: test/gtest/uct/ib/test_devx.cc (hunk 13 above)
Comment (yosefe) [reply test/gtest/uct/ib/test_devx.cc:106]: i think it's not important to check if umr objects are initialized, but just try to use them (it should fail if they are not initialized)

File: test/gtest/uct/ib/test_devx.cc (hunk 14 above)
Comment (yosefe) [reply test/gtest/uct/ib/test_devx.cc:112]: > * Validate import of non-UMR mkey - to make sure that UMR workflow is NOT used, and UMR alias hash map remains unallocated

Wouldn't the import fail anyway if UMR workflow is used for non-UMR mkey?

File: test/gtest/uct/ib/test_devx.cc (hunk 15 above)
Comment (yosefe) [reply test/gtest/uct/ib/test_devx.cc:120]: Maybe we can check that UMR is used and mpool keys are reused by time measurement or counting FW commands. We expect that we could import mkeys fast and without FW command.

File: test/gtest/uct/ib/test_devx.cc (hunk 20 above)
Comment (yosefe) [reply test/gtest/uct/ib/test_devx.cc:214]: > 1. Add configuration e.g. `UCX_UMR_ENABLE` like you said. IMO this is the least desirable option, because there is no use case in production workflow for this option (right?), so we add it in production code purely for unit testing purposes, which seems wrong.

Actually it can be useful to add option to enable/disable this feature in case we find SW/HE issues with this flow. Since it can be disabled without hurting functionality.

File: test/gtest/uct/ib/test_devx.cc (hunk 9 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:71]: Because those "fake-xgvmi" devices are real devices that we support. For instance, when I run my test on rock, I see that both mlx5_0 and mlx5_2 (MT4129 == ConnectX7?) are examples of those devices. I've no idea why they advertise XGVMI capability. But since we **can** run export mkey on those devices, I prefer to guarantee that it works as expected (= not using UMR approach).

File: test/gtest/uct/ib/test_devx.cc (hunk 14 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:112]: No, it does not fail at all. It can successfully create an alias for non-UMR keys and store it inside the hash map

mem_attach supports both UMR mkeys and KSM/UMEM generated mkeys. That's why we distinguish between them by checking the mkey tag. So after importing a key I want to make sure that UMR approach was not triggered for non-UMR key

File: test/gtest/uct/ib/test_devx.cc (hunk 15 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:120]: Time measurement is not reliable method I think. Also it's error prone, because overall performance depends on too many different calls.
First, I don't even know how we can count these FW commands?
Second, counting FW commands - it may be even less reliable in this case, because we want to differentiate between UMR and non=UMR workflow. But both workflows call FW commands (for creating mkey and then allow XGVMI).

File: test/gtest/uct/ib/test_devx.cc (hunk 20 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:214]: ok, adding an option then

File: test/gtest/uct/ib/test_devx.cc (hunk 14 above)
Comment (yosefe) [reply test/gtest/uct/ib/test_devx.cc:112]: why is it a problem that  UMR approach is triggered for non-UMR key?

File: test/gtest/uct/ib/test_devx.cc (hunk 15 above)
Comment (yosefe) [reply test/gtest/uct/ib/test_devx.cc:120]: UMR should not do FW flow in most cases, this is the goal of using it.. so let's test it

File: test/gtest/uct/ib/test_devx.cc (hunk 16 above)
Comment (yosefe) [reply test/gtest/uct/ib/test_devx.cc:152]: would ASAN catch it if we had a memory leak?

File: test/gtest/uct/ib/test_devx.cc (hunk 14 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:112]: This would be a real problem (and for some time during development it was really the case)
With UMR approach we store all the received aliases in the hash map, and we destroy them only at the end of process lifetime. So if we do this for non-UMR key - we would store alias to non-UMR key forever. This is not a classical memory leak, because we destroy those objects at the end, but this is abusive memory usage for no reason.

File: test/gtest/uct/ib/test_devx.cc (hunk 15 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:120]: True, UMR aims to minimize the number of FW calls. We could check that FW is not called in `export_mkey_xgvmi_reuse` test, it would be a nice test. The only problem is that I don't know how to count those FW calls... Do you have a clue?

We could use mocking framework for that, but this is a BIG change

File: test/gtest/uct/ib/test_devx.cc (hunk 16 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:152]: Well, if we had a true memory leak, it must be catch by LSAN.
Here however I don't check for a memory leak. All the objects stored in the mkey pool or alias hash map will be destroyed at the end of the process, so no memory leak will be detected in this particular case. I just validate the invariant: pool must not grow up on export.
Again, I don't consider this check as super important, so we can remove it if you dislike it

File: test/gtest/uct/ib/test_devx.cc (hunk 15 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:133]: Added perf test, as we agreed, measuring CPU cycles of export & import

File: test/gtest/uct/ib/test_devx.cc (hunk 10 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc]: ok, done

File: test/gtest/uct/ib/test_devx.cc (hunk 13 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:109]: well, the most important use case here is to check that UMR objects are NOT initialized (when they should not be)

File: test/gtest/uct/ib/test_devx.cc [hunk 21]
```diff
@@ -49,3 +49,235 @@ UCS_TEST_P(test_devx, dbrec)
...
+    ASSERT_EQ(MEMH_COUNT, ucs_list_length(&md()->umr.mkey_pool));
+
+    /* Now reuse the keys for different handles of different size */
+    ucs_time_t export_reuse_time, import_reuse_time;
+    perf_export_import_mkey(MEMH_COUNT, 2048, export_reuse_time,
+                            import_reuse_time);
+
+    /* Reusing existing UMR mkeys is normally 200x faster than creating new
+     * ones, given jitter we check that it's at least 50 times faster */
+    ASSERT_GT(export_create_time / 50, export_reuse_time);
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:273]: 1. IMO we can reduce it to 10 times, or check vs. some known time content
2. inverse order since export_reuse_time is what we test:
EXPECT_LT(export_reuse_time, export_create_time / 10)

File: test/gtest/uct/ib/test_devx.cc [hunk 22]
```diff
@@ -49,3 +49,235 @@ UCS_TEST_P(test_devx, dbrec)
...
+    ASSERT_GT(import_create_time / 50, import_reuse_time);
+    UCS_TEST_MESSAGE << "Export create time: " << export_create_time;
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:275]: IMO we can print it in a more compact way + convert to milliseconds

Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:275]: Ok, now it's in this format:
```
[     INFO ] Create 1000 UMR mkeys, export: 397.802 ms, import: 215.757 ms       
[     INFO ] Reuse 1000 UMR mkeys, export: 1.784 ms, import: 0.421 ms                                            
```

File: test/gtest/uct/ib/test_devx.cc [hunk 23]
```diff
@@ -49,3 +49,235 @@ UCS_TEST_P(test_devx, dbrec)
...
+
+UCS_TEST_P(test_devx_umr_mkey, import_mkey_no_umr, "IB_XGVMI_UMR_ENABLE=n")
+{
+    if (!check_xgvmi()) {
+        UCS_TEST_SKIP_R("XGVMI capability is not supported");
+    }
+
+    /* Export any mkey */
+    uct_ib_mlx5_devx_mem_t *exported_memh = create_memh(1024);
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), exported_memh));
```
[the remark this answers is not in the document]
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:208]: I think it's redundant, because export workflow we check already in export_mkey_no_xgvmi test

File: test/gtest/uct/ib/test_devx.cc [hunk 24]
```diff
@@ -49,3 +49,234 @@ UCS_TEST_P(test_devx, dbrec)
...
+    perf_export_import_mkey(MEMH_COUNT, 2048, export_reuse_time,
+                            import_reuse_time);
+
+    /* Reusing existing UMR mkeys is normally 200x faster than creating new
+     * ones, given jitter we check that it's at least 10 times faster */
+    EXPECT_LT(export_reuse_time, export_create_time / 10);
+    EXPECT_LT(export_reuse_time, export_create_time / 10);
+
+    char buf[1024] = {};
+    sprintf(buf, "Create %zu UMR mkeys, export: %.3f ms, import: %.3f ms",
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:270]: snprintf / ucs_snprintf_safe

File: test/gtest/uct/ib/test_devx.cc [hunk 25]
```diff
@@ -49,3 +49,234 @@ UCS_TEST_P(test_devx, dbrec)
...
+            MEMH_COUNT, ucs_time_to_msec(export_create_time),
+            ucs_time_to_msec(import_create_time));
+    UCS_TEST_MESSAGE << buf;
+    sprintf(buf, "Reuse %zu UMR mkeys, export: %.3f ms, import: %.3f ms",
```
Comment (yosefe) [inline test/gtest/uct/ib/test_devx.cc:274]: snprintf / ucs_snprintf_safe

File: test/gtest/uct/ib/test_devx.cc (hunk 24 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:270]: sure

File: test/gtest/uct/ib/test_devx.cc (hunk 25 above)
Comment (iyastreb) [reply test/gtest/uct/ib/test_devx.cc:274]: done

**Review by yosefe — APPROVED** [review]
